waku/v2/api/publish/message_check.go

```diff
@@ -0,0 +1,198 @@
+package publish
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/pb"
+	"github.com/waku-org/go-waku/waku/v2/protocol/store"
+	"github.com/waku-org/go-waku/waku/v2/timesource"
+	"go.uber.org/zap"
+)
+
+const maxHashQueryLength = 100
+const hashQueryInterval = 3 * time.Second
+const messageSentPeriod = 3    // in seconds
+const messageExpiredPerid = 10 // in seconds
+
+type MessageSentCheck struct {
+	messageIDs         map[string]map[common.Hash]uint32
+	messageIDsMu       sync.RWMutex
+	storePeerID        peer.ID
+	MessageStoredChan  chan common.Hash
+	MessageExpiredChan chan common.Hash
+	ctx                context.Context
+	store              *store.WakuStore
+	timesource         timesource.Timesource
+	logger             *zap.Logger
+}
+
+func NewMessageSentCheck(ctx context.Context, store *store.WakuStore, timesource timesource.Timesource, logger *zap.Logger) *MessageSentCheck {
+	return &MessageSentCheck{
+		messageIDs:         make(map[string]map[common.Hash]uint32),
+		messageIDsMu:       sync.RWMutex{},
+		MessageStoredChan:  make(chan common.Hash, 1000),
+		MessageExpiredChan: make(chan common.Hash, 1000),
+		ctx:                ctx,
+		store:              store,
+		timesource:         timesource,
+		logger:             logger,
+	}
+}
+
+func (m *MessageSentCheck) Add(topic string, messageID common.Hash, sentTime uint32) {
```

Comment:
Does this have to be invoked by user? 
Maybe we should invoke this for relay and lightpush if this feature is enabled? 
Wondering if this can be done at api layer for relay and lightpush i.e have a send wrapper which would internally decide which protocol to use and at the same time add to this queue if this feature is enabled.

WDYT @richard-ramos

Reply:
It does not sound like a bad idea at all, specially since I think the parameters for this `Add` function could be obtained from the envelope

Reply:
Yeah, sounds great!
We can add a sent method to replace lightpush.publish and relay.publish with the added message sent check.
I will try such refactor in new PRs.

Reply:
Can we have an issue tracking so that we don't miss this?

Reply:
via https://github.com/waku-org/go-waku/issues/1181